sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala

```diff
@@ -45,7 +45,8 @@ import org.apache.spark.util.Utils
 class QueryExecution(
     val sparkSession: SparkSession,
     val logical: LogicalPlan,
-    val tracker: QueryPlanningTracker = new QueryPlanningTracker) {
+    val tracker: QueryPlanningTracker = new QueryPlanningTracker,
+    val alreadyOptimized: Boolean = false) {
```

Comment:
We could probably use this for subqueries too.

Reply:
@maryannxue, I plan to add subquery reuse across recursive iterations with the help of this: https://github.com/apache/spark/pull/25644